File: data/derived_cot/rq1_traces/translation/qwen/py_to_cpp/code_output/Warehouse.cpp
#include <unordered_map>
#include <optional>
#include <string>

struct Product {
    std::string name;
    int quantity;
};

struct Order {
    int product_id;
    int quantity;
    std::string status;
};

class Warehouse {
private:
    std::unordered_map<int, Product> inventory;
    std::unordered_map<int, Order> orders;

public:
    Warehouse() = default;

    void add_product(int product_id, const std::string& name, int quantity) {
        if (inventory.find(product_id) == inventory.end()) {
            inventory[product_id] = {name, quantity};
        } else {
            inventory[product_id].quantity += quantity;
        }
    }

    void update_product_quantity(int product_id, int quantity) {
        if (inventory.find(product_id) != inventory.end()) {
            inventory[product_id].quantity += quantity;
        }
    }

    std::optional<int> get_product_quantity(int product_id) {
        auto it = inventory.find(product_id);
        if (it != inventory.end()) {
            return it->second.quantity;
        }
        return std::nullopt;
    }

    bool create_order(int order_id, int product_id, int quantity) {
        auto quantity_opt = get_product_quantity(product_id);
        if (!quantity_opt.has_value()) {
            return false;
        }
        if (*quantity_opt < quantity) {
            return false;
        }
        update_product_quantity(product_id, -quantity);
        orders[order_id] = {product_id, quantity, "Shipped"};
        return true;
    }

    bool change_order_status(int order_id, const std::string& status) {
        auto it = orders.find(order_id);
        if (it != orders.end()) {
            it->second.status = status;
            return true;
        }
        return false;
    }

    std::optional<std::string> track_order(int order_id) {
        auto it = orders.find(order_id);
        if (it != orders.end()) {
            return it->second.status;
        }
        return std::nullopt;
    }
};

File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/cpp_to_py/Warehouse.py
import types

class Warehouse:
    def __init__(self):
        self.inventory_ = {}
        self.orders_ = {}

    def add_product(self, product_id, name, quantity):
        if product_id not in self.inventory_:
            self.inventory_[product_id] = {"name": name, "quantity": quantity}
        else:
            current_quantity = self.inventory_[product_id]["quantity"]
            self.inventory_[product_id]["quantity"] = current_quantity + quantity

    def update_product_quantity(self, product_id, quantity):
        if product_id in self.inventory_:
            current_quantity = self.inventory_[product_id]["quantity"]
            self.inventory_[product_id]["quantity"] = current_quantity + quantity

    def get_product_quantity(self, product_id):
        if product_id in self.inventory_:
            return self.inventory_[product_id]["quantity"]
        return 0

    def create_order(self, order_id, product_id, quantity):
        available_quantity = self.get_product_quantity(product_id)
        if available_quantity >= quantity:
            self.update_product_quantity(product_id, -quantity)
            order_entry = {
                "product_id": str(product_id),
                "quantity": str(quantity),
                "status": "Shipped"
            }
            self.orders_[order_id] = order_entry
            return True
        return False

    def change_order_status(self, order_id, status):
        if order_id in self.orders_:
            self.orders_[order_id]["status"] = status
            return True
        return False

    def track_order(self, order_id):
        if order_id in self.orders_:
            return self.orders_[order_id]["status"]
        return ""

    def orders(self):
        return types.MappingProxyType(self.orders_)

    def inventory(self):
        return types.MappingProxyType(self.inventory_)